Add documentation for parallel_phase feature (#1646)
uxlfoundation/oneTBB: doc/main/reference/parallel_phase_for_task_arena.rst

Reviewer: How about referencing the corresponding RFC document from this file?

Author: Done.

Reviewer: The links to the RFC and to the specification (are there other?) are not permanent and therefore will become obsolete someday. I suggest to put permanent links so that they stay relevant as long as possible.

.. _parallel_phase_for_task_arena:

``parallel_phase`` Interface for Task Arena
====================================================================

.. note::

>     To enable this feature, set ``TBB_PREVIEW_PARALLEL_PHASE`` macro to 1.

Author: Added the link.

with the following API:

* Adds the ``leave_policy`` enumeration class to ``task_arena``.
* Adds ``leave_policy`` as the last parameter in ``task_arena`` constructors and ``task_arena::initialize`` methods.
  This allows you to inform the scheduler about the preferred policy for worker threads
  when they are about to leave `task_arena` due to a lack of available work.
* Adds new ``start_parallel_phase`` and ``end_parallel_phase`` interfaces to the ``task_arena`` class
  and the ``this_task_arena`` namespace. These interfaces work as hints to the scheduler to mark the start and end
  of parallel work in the arena, enabling different worker thread retention policies.
* Adds the RAII class ``scoped_parallel_phase`` to ``task_arena``.

API
***

Header
------

.. code:: cpp

    #define TBB_PREVIEW_PARALLEL_PHASE 1
    #include <oneapi/tbb/task_arena.h>

Synopsis
--------

.. code:: cpp

    namespace oneapi {
        namespace tbb {

            class task_arena {
            public:

                enum class leave_policy : /* unspecified type */ {
                    automatic = /* unspecifed */,
                    fast = /* unspecifed */,
                };

>                 task_arena(int max_concurrency = automatic, unsigned reserved_for_masters = 1,

Reviewer 2: What about the copy constructor? It copies settings from the other arena, including the leave policy. I think this behavior should be stated somewhere?

Author: Since the API for the copy constructor is not changed, I think the only way to state that is to copy description from the spec and add info about the leave policy.

Author: Listed copy constructor in member functions.

                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);

                class scoped_parallel_phase {
                public:
                    scoped_parallel_phase(task_arena& ta, bool with_fast_leave = false);
                };
            }; // class task_arena

            namespace this_task_arena {
                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);
            } // namespace this_task_arena

        } // namespace tbb
    } // namespace oneapi

Member Types
----------------

.. cpp:enum:: leave_policy::automatic

When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has

> default policy for worker threads.

Reviewer 2: Perhaps, "has the default (possibly system specific) policy for how quickly worker threads leave the arena when there is no more work available in the arena and when the arena is not in a parallel phase."

> default policy for worker threads.

Author: Done.

.. note:: Worker threads in ``task_arena`` might be retained based on internal heuristics.

.. cpp:enum:: leave_policy::fast

When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has
policy to not retain worker threads in ``task_arena``.

> 

Author: Applied suggested wording. I also added some notes on behavior when threads re-join the arena with no active parallel phases.

.. cpp:class:: scoped_parallel_phase

The RAII class to map a parallel phase to a code scope.

.. cpp:function:: scoped_parallel_phase::scoped_parallel_phatse(task_arena& ta, bool with_fast_leave = false)

Constructs a ``scoped_parallel_phase`` object that starts a parallel phase in the specified ``task_arena``.

> If ``with_fast_leave`` is ``true``, the worker threads leave policy is temporarily set to ``fast``.

Author: Added note about the effect of this setting with arena initialized with fast leave policy.

.. _parallel_phase_for_task_arena:

parallel_phase interface to hint start and end of work in task arena
====================================================================

.. note::
    To enable this feature, set ``TBB_PREVIEW_PARALLEL_PHASE`` macro to 1.

.. contents::
    :local:
    :depth: 1

Description
***********

> This feature extends the `tbb::task_arena specification <https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/task_scheduler/task_arena/task_arena_cls>`_

Reviewer 3: make sense, thanks for explaining!

with the following API:

> * Add enumeration class ``leave_policy`` to ``task_arena``.

Author: Done.

* Add ``leave_policy`` as the last parameter to ``task_arena`` constructors and ``task_arena::initialize`` methods.

>   This allows to choose retention policy for worker threads in case when no more parallel work in the arena. 

Author: I am not sure how to better state this but at the same time to not overwhelm documentation with lots of details. The idea is that the feature allows users to tell the scheduler, what policy to use when TBB's worker threads **are going to leave** `task_arena`.

>   This allows to choose retention policy for worker threads in case when no more parallel work in the arena. 

Reviewer 3: maybe smth like: This feature allows you to inform the scheduler about the preferred policy for worker threads when they are about to leave task_arena due to a lack of available work.

>   This allows to choose retention policy for worker threads in case when no more parallel work in the arena. 

Author: Updated.

                void initialize(int max_concurrency, unsigned reserved_for_masters = 1,
                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);

                class scoped_parallel_phase {
                public:
                    scoped_parallel_phase(task_arena& ta, bool with_fast_leave = false);
                };
            }; // class task_arena

            namespace this_task_arena {
                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);
            } // namespace this_task_arena

        } // namespace tbb
    } // namespace oneapi

Member types and constants
----------------

.. cpp:enum:: leave_policy::automatic

>     When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has

Author: I tried to say here that "if `leave_policy::automatic` constant is passed to the `task_arena` constructor or `initialize` method, then etc and etc".

>     When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has

Reviewer 3: got it!

    default policy for worker threads.

    . note::

        Worker threads might be retained in ``task_arena`` based on internal heuristics.

.. cpp:enum:: leave_policy::fast

>     When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has

Author: Same logic as for leave_policy::automatic comment.

    Indicates the point from which the scheduler can use a hint and keep threads in the arena for longer.

    . note::

        This function may serve as a warm-up hint to the scheduler as well: the scheduler may wake
        up worker threads in advance.

.. cpp:function:: void end_parallel_phase(bool with_fast_leave = false)

    Indicates the point from which the scheduler may drop the hint and no longer retain threads in the arena.
    If ``with_fast_leave`` is true, worker threads' leave policy is temporarily overridden to ``fast``.

Functions
---------

.. cpp:function:: void this_task_arena::start_parallel_phase()

    Indicate the start of parallel phase in the current ``task_arena``.

.. cpp:function:: void this_task_arena::end_parallel_phase(bool with_fast_leave = false)

    Indicate the end of parallel phase in the current ``task_arena``.
    If ``with_fast_leave`` is true, worker threads' leave policy is temporarily overridden to ``fast``.

Example
*******

.. literalinclude:: .examples/parallel_phase_example.cpp
   :language: c++
   :start-after: /*begin_parallel_phase_example*/
   :end-before: /*end_parallel_phase_example*/

> In this example, ``task_arena`` is created with ``leave_policy::fast`` which means that

Reviewer 3: ```suggestion
In this example, ``task_arena`` is created with ``leave_policy::fast``. It means that
```

> it is expected that worker threads shouldn't be retained in ``task_arena`` when parallel work is done.

Reviewer 3: ```suggestion
worker threads are not expected to remain in ``task_arena`` once parallel work is completed.
```

> However, there is a sequence of parallel work (initializing and sorting data) interceded by serial work (prefix sum).

Reviewer 3: ```suggestion
However, the workflow includes a sequence of parallel work (initializing and sorting data) interceded by serial work (prefix sum).
```

> To hint the start and end of parallel work, ``scoped_parallel_phase`` is used. This allows to hint the scheduler

Reviewer 3: ```suggestion
To hint the start and end of parallel work, ``scoped_parallel_phase`` is used. This provides a hint to the scheduler
```

                void initialize(int max_concurrency, unsigned reserved_for_masters = 1,
                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
                                priority a_priority = priority::normal,
                                leave_policy a_leave_policy = leave_policy::automatic);

                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);

                class scoped_parallel_phase {
                public:
                    scoped_parallel_phase(task_arena& ta, bool with_fast_leave = false);
                };
            }; // class task_arena

            namespace this_task_arena {
                void start_parallel_phase();
                void end_parallel_phase(bool with_fast_leave = false);
            } // namespace this_task_arena

        } // namespace tbb
    } // namespace oneapi

Member Types
----------------

.. cpp:enum:: leave_policy::automatic

>     When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has

Reviewer 3: ```suggestion
When passed to a constructor or the ``initialize`` method, the initialized ``task_arena`` has
```

.. cpp:function:: scoped_parallel_phase::scoped_parallel_phatse(task_arena& ta, bool with_fast_leave = false)

Constructs a ``scoped_parallel_phase`` object that starts a parallel phase in the specified ``task_arena``.
If ``with_fast_leave`` is ``true``, the worker threads leave policy is temporarily set to ``fast``.

Member Functions
----------------

.. cpp:function:: void start_parallel_phase()

Indicates a point from where the scheduler can use a hint to keep threads in the arena for longer.

.. note:: This function can also be a warm-up hint for the scheduler. It allows the scheduler to wake up worker threads in advance.

       
        

.. cpp:function:: void end_parallel_phase(bool with_fast_leave = false)

Indicates the point when the scheduler may drop a hint and no longer retain threads in the arena.
If ``with_fast_leave`` is ``true``, worker threads leave policy is temporarily set to ``fast``.

Functions
---------

.. cpp:function:: void this_task_arena::start_parallel_phase()

Indicates the start of the parallel phase in the current ``task_arena``.

.. cpp:function:: void this_task_arena::end_parallel_phase(bool with_fast_leave = false)

> Indicate the end of the parallel phase in the current ``task_arena``.

Reviewer 3: ```suggestion
Indicates the end of the parallel phase in the current ``task_arena``.
```
